### PR TITLE
feat: cuvs acceleration for gpu k-means

### PR DESCRIPTION
**Pull request by jacketsj**
We currently have a pytorch-based k-means implementation for computing IVF centroids. This PR accelerates it with cuVS.
This uses a tradeoff of faster iterations/less score improvement per iteration.

By default, this is off, since it's primarily useful for very large datasets where large centroid counts are applicable.

Benchmarking (classic k-means scoring):
- k=16384 clusters
- text2image-10M base set (10M image embeddings, 200 dimensions, float32, cosine distance)

Results: Slightly better score @ ~1.5x faster. Speedup gets better with more centroids.

<details>

<summary>Easy test script & outputs</summary>

```py
import numpy as np
from lance.cuvs.kmeans import KMeans as KMeansVS
from lance.torch.kmeans import KMeans
import lance
import time
# Note: This kind of approach performs quite poorly on random data (see https://arxiv.org/abs/2405.18680), so it's only worth testing on a real dataset
ds = lance.dataset("path/to/text2image-dataset") # can also use other medium~large datasets
data = np.stack(ds.to_table()["vector"].to_numpy())
max_iters_base = 10
max_iters_cuvs = 12 # iters using cuvs are much faster, but slightly less precise
metric = "cosine"

cuvs_start_time = time.time()
kmeans_cuvs = KMeansVS(
    CLUSTERS,
    metric=metric,
    max_iters=max_iters_cuvs,
    seed=0,
)
kmeans_cuvs.fit(data)
cuvs_end_time = time.time()

base_start_time = time.time()
kmeans = KMeans(
    CLUSTERS,
    metric=metric,
    max_iters=max_iters_base,
    seed=0,
)
kmeans.fit(data)
base_end_time = time.time()
print(f"score after {max_iters_cuvs} iters of kmeans_cuvs better than {max_iters_base} iters of kmeans by {kmeans.total_distance - kmeans_cuvs.total_distance}")
base_time = base_end_time-base_start_time
cuvs_time = cuvs_end_time-cuvs_start_time
print(f"time to run kmeans: {base_time}s. time to run kmeans_cuvs: {cuvs_time} (speedup: {base_time/cuvs_time}x)")
```
Output:
```
score after 12 iters of kmeans_cuvs better than 10 iters of kmeans by 5905.7138671875
time to run kmeans: 86.69116258621216s. time to run kmeans_cuvs: 56.66267776489258 (speedup: 1.5299517425899842x)
```
</details>

Additionally, a new "accelerator" choice has been added: "cuvs". This requires one of the added optional dependencies (cuvs-py3X, X in {9,10,11}). This can replace the two routines for which we already have cuda acceleration: IVF model training (Lloyd's algorithm) and IVF assignments. At sufficiently large centroid counts, this can significantly accelerate these steps, resulting in better e2e time. See below:

results_static_20240906_132311_plot_dataset_sift1m_k_10
Although these plots are near-identical, the "cuvs" accelerated variation took ~18.1s to build e2e, while the "cuda" accelerated variation took ~24.4s.

This speedup persists on larger datasets, although I was mistaken in that PQ assignments are a bigger bottleneck as the dataset gets larger (thanks to some improvements I did not see), so this is not the bottleneck step. The next step after this PR will be to accelerate PQ with both cuda and cuvs.

### CODE REVIEW

File: python/python/lance/torch/data.py [hunk 1]
```diff
@@ -51,10 +51,10 @@ def _to_tensor(
             pa.types.is_floating(arr.type.value_type)
             or pa.types.is_integer(arr.type.value_type)
         ):
-            np_tensor = arr.values.to_numpy(zero_copy_only=True).reshape(
-                -1, arr.type.list_size
-            )
-            tensor = torch.from_numpy(np_tensor)
+            np_arrs = arr.to_numpy(zero_copy_only=False)
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/data.py:54]: cc: @eddyxu there are weird conditions that trigger this thing. I have no idea what are those

File: python/python/lance/torch/kmeans.py [hunk 2]
```diff
@@ -72,13 +88,26 @@ def __init__(
                 f"Only l2/cosine/dot is supported as metric type, got: {metric}"
             )
 
+        if self.use_cuvs and self.metric == "dot":
+            logging.warning(
+                'Kmeans::__init__: metric == "dot" is incompatible" \
+                " with use_cuvs == True, disabling cuvs...'
+            )
+            self.use_cuvs = False
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/kmeans.py:96]: just raise an error

File: python/python/lance/torch/kmeans.py [hunk 3]
```diff
@@ -72,13 +88,26 @@ def __init__(
...
         self.init = init
         self.device = preferred_device(device)
         self.tolerance = tolerance
         self.seed = seed
 
+        if self.device.type != "cuda" and self.use_cuvs:
+            logging.warning(
+                "Kmeans::__init__: cuda is not enabled/available, disabling cuvs..."
+            )
+            self.use_cuvs = False
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/kmeans.py:109]: same, just raise

File: python/pyproject.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [project]
 name = "pylance"
-dependencies = ["pyarrow>=12", "numpy>=1.22,<2"]
+dependencies = ["pyarrow>=12", "numpy>=1.22,<2", "cuvs-cu12", "pylibraft-cu12"]
```
Comment (eddyxu) [inline python/pyproject.toml:3]: lets not put this into required dependencies?

File: python/python/lance/torch/data.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py:54]: This line significantly slow down torch training. 

@jacketsj could you investigate a bit more what data cause this line to fail?

And how much slow down it is due to this line change to pytorch kmeans?

File: python/python/lance/torch/kmeans.py [hunk 5]
```diff
@@ -43,6 +51,8 @@ class KMeans:
     device: str, optional
         The device to run the PyTorch algorithms. Default we will pick
         the most performant device on the host. See `lance.torch.preferred_device()`
+    use_cuvs: bool
```
Comment (eddyxu) [inline python/python/lance/torch/kmeans.py:54]: can we detect this automatically? i.e., `devcie == cuda AND num_partitons > blablabla`

File: python/pyproject.toml [hunk 6]
```diff
@@ -64,6 +64,7 @@ tests = [
 dev = ["ruff==0.4.1"]
 benchmarks = ["pytest-benchmark"]
 torch = ["torch"]
+cuvs = ["cuvs-cu12", "pylibraft-cu12"]
```
Comment (eddyxu) [inline python/pyproject.toml:67]: do we need to limit the platform to linux only? Will it fail on mac / windows?

File: python/python/lance/torch/kmeans.py [hunk 7]
```diff
@@ -271,6 +330,30 @@ def _transform(
         if self.metric == "cosine":
             data = torch.nn.functional.normalize(data)
 
+        if self.use_cuvs:
+            search_time_start = time.time()
+            device = torch.device("cuda")
+            out_idx = device_ndarray.empty((data.shape[0], 1), dtype="uint32")
+            out_dist = device_ndarray.empty((data.shape[0], 1), dtype="float32")
+            search_params = cagra.SearchParams(itopk_size=self.itopk_size)
+            cagra.search(
```
Comment (eddyxu) [inline python/python/lance/torch/kmeans.py:339]: is inference faster than torch significantly?

File: python/python/lance/torch/kmeans.py [hunk 8]
```diff
@@ -263,6 +304,24 @@ def _fit_once(
         )
         return total_dist
 
+    def rebuild_index(self):
+        if self.use_cuvs:
+            rebuild_time_start = time.time()
```
Comment (eddyxu) [inline python/python/lance/torch/kmeans.py:309]: would it be easy and clean to have a `cuvs.KMeans`, don't need to share with this Torch code. Mixing cuvs with torch is not maintainable.

File: python/python/lance/torch/data.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py:54]: this `zero_copy_only=True` is part of fix to improve GPU inference by 2-3x times end-to-end, where we measured as one of the slowest step during large scale indexing.

If we can not figure out exactly what cause the failure, at least lets do:
```
try:
  tensor = torch.from_numpy(arr.values.to_numpy(zero_copy_only=True)...)
except SomeException:
  tensor = tensor.from_array(np.stack(arr.to_numpy(zero_copy_only=False)))
```

File: python/python/lance/torch/kmeans.py (hunk 7 above)
Comment (eddyxu) [reply python/python/lance/torch/kmeans.py:339]: Is this algorithm search-able on CPU? i.e., can we run indexing job on GPU but lancedb serving on CPU?

File: python/python/lance/torch/kmeans.py (hunk 8 above)
Comment (jacketsj) [reply python/python/lance/torch/kmeans.py]: Done

File: python/python/lance/torch/kmeans.py (hunk 7 above)
Comment (jacketsj) [reply python/python/lance/torch/kmeans.py]: This doesn't change the query path, so queries are still fully CPU-based.
Batch inference is significantly faster than brute-force via torch for sufficiently large numbers of centroids, at a small cost to accuracy/recall (I'll be putting up a large e2e test result before asking for re-review).

File: python/pyproject.toml (hunk 6 above)
Comment (jacketsj) [reply python/pyproject.toml]: I don't have a windows/mac machine to test on, but this does require a cuda device. It may be worth testing on windows in the future.

File: python/python/lance/torch/kmeans.py (hunk 5 above)
Comment (jacketsj) [reply python/python/lance/torch/kmeans.py]: Unfortunately this adds another dataset-dependent parameter with no obvious threshold/choice.
I set up a new "accelerator" type for this. We could add automatic accelerator choices in the future though (just as decent untuned defaults).

File: python/pyproject.toml (hunk 4 above)
Comment (jacketsj) [reply python/pyproject.toml]: Done

File: python/python/lance/torch/kmeans.py (hunk 3 above)
Comment (jacketsj) [reply python/python/lance/torch/kmeans.py]: Done

File: python/python/lance/torch/kmeans.py (hunk 2 above)
Comment (jacketsj) [reply python/python/lance/torch/kmeans.py]: Done

File: python/python/lance/torch/data.py (hunk 1 above)
Comment (jacketsj) [reply python/python/lance/torch/data.py]: Moved this code into #2824 after investigating.
